tsl/src/fdw/relinfo.c

```diff
@@ -308,74 +245,97 @@ estimate_tuples_and_pages_using_shared_buffers(PlannerInfo *root, Hypertable *ht
+estimate_chunk_size(PlannerInfo *root, RelOptInfo *chunk_rel)
 {
-	int parent_relid;
-	RangeTblEntry *hyper_rte;
-	Cache *hcache;
-	Hypertable *ht;
-	double chunk_fillfactor;
-	RangeTblEntry *chunk_rte;
-	Chunk *chunk;
-	Hyperspace *hyperspace;
-	RelEstimates *estimates;
-
-	Assert(rel->tuples <= 0);
-	Assert(rel->pages == 0);
-
-	/* In some cases (e.g., UPDATE stmt) top_parent_relids is not set so the best
-		we can do is using shared buffers size without partitioning information.
-	   Since updates are not something we generaly optimize for this should be fine. */
-	if (rel->top_parent_relids == NULL)
+	const int parent_relid = bms_next_member(chunk_rel->top_parent_relids, -1);
+	if (parent_relid < 0)
 	{
-		estimates =
-			estimate_tuples_and_pages_using_shared_buffers(root, NULL, rel->reltarget->width);
-		set_rel_estimates(rel, estimates);
+		/*
+		 * In some cases (e.g., UPDATE stmt) top_parent_relids is not set so the
+		 * best we can do is using shared buffers size without partitioning
+		 * information. Since updates are not something we generaly optimize
+		 * for, this should be fine.
+		 */
+		if (chunk_rel->pages == 0 && chunk_rel->tuples <= 0)
+		{
+			estimate_tuples_and_pages_using_shared_buffers(root, NULL, chunk_rel);
+		}
 		return;
 	}
 
-	parent_relid = bms_next_member(rel->top_parent_relids, -1);
-	hyper_rte = planner_rt_fetch(parent_relid, root);
-	hcache = ts_hypertable_cache_pin();
-	ht = ts_hypertable_cache_get_entry(hcache, hyper_rte->relid, CACHE_FLAG_NONE);
-	hyperspace = ht->space;
-	chunk_rte = planner_rt_fetch(rel->relid, root);
-	chunk = ts_chunk_get_by_relid(chunk_rte->relid, true);
-
-	/* Let's first try figuring out number of tuples/pages using stats from previous chunks,
-	otherwise make an estimation based on shared buffers size */
-	estimates = estimate_tuples_and_pages_using_prev_chunks(root, hyperspace, chunk);
-	if (estimates->tuples <= 0 || estimates->pages == 0)
-		estimates = estimate_tuples_and_pages_using_shared_buffers(root, ht, rel->reltarget->width);
-
-	chunk_fillfactor = estimate_chunk_fillfactor(chunk, hyperspace);
-	/* adjust tuples/pages using chunk_fillfactor */
-	rel_estimates_apply_fillfactor(estimates, chunk_fillfactor);
-	set_rel_estimates(rel, estimates);
+	RelOptInfo *parent_info = root->simple_rel_array[parent_relid];
+	TsFdwRelInfo *parent_private = fdw_relinfo_get(parent_info);
+	RangeTblEntry *parent_rte = planner_rt_fetch(parent_relid, root);
+	Cache *hcache = ts_hypertable_cache_pin();
+	Hypertable *ht = ts_hypertable_cache_get_entry(hcache, parent_rte->relid, CACHE_FLAG_NONE);
+	Hyperspace *hyperspace = ht->space;
+	RangeTblEntry *chunk_rte = planner_rt_fetch(chunk_rel->relid, root);
+	Chunk *chunk = ts_chunk_get_by_relid(chunk_rte->relid, true /* fail_if_not_found */);
+
+	const double fillfactor = estimate_chunk_fillfactor(chunk, hyperspace);
+	const bool have_chunk_statistics = chunk_rel->pages != 0 || chunk_rel->tuples > 0;
```

Comment:
Should we check that we have both pages and tuples set? Or assert it? (Can you have tuples stored without a page?)

```suggestion
	const bool have_chunk_statistics = chunk_rel->pages != 0 && chunk_rel->tuples > 0;
```